### PR TITLE
Fix #2399: Cannot change the type of the Service from ClusterIP to ExternalName

### LINKED ISSUE #2399: Cannot change the type of the Service from ClusterIP to ExternalName
The current implementation ignores changes to the ClusterIP field (related to: [https://github.com/fabric8io/kubernetes-client/issues/1925](https://github.com/fabric8io/kubernetes-client/issues/1925)), it copies the old value from the server entity and does not include it in the PATCH call.

When editing the 'type' field from ClusterIP to ExternalName the API Servers enforces the ClusterIP field to be empty making the edit impossible using the kubernetes-client.

Validation message: Message: Service "elasticsearch-logging" is invalid: spec.clusterIP: Forbidden: must be empty for ExternalName services. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=spec.clusterIP, message=Forbidden: must be empty for ExternalName services, reason=FieldValueForbidden, additionalProperties={})], group=null, kind=Service, name=elasticsearch-logging, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Service "elasticsearch-logging" is invalid: spec.clusterIP: Forbidden: must be empty for ExternalName services, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

Fix #2399 

Removes overridden replace() method in `ServiceOperationsImpl`

## Type of change

 - [X] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [X] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Review by manusa — APPROVED** [review]

**Comment by manusa** [comment]
[merge]

**Review by dev-gaur — CHANGES_REQUESTED** [review]

File: kubernetes-itests/src/test/java/io/fabric8/kubernetes/ServiceIT.java [hunk 1]
```diff
@@ -94,6 +96,135 @@ public void delete() {
...
 
+  @Test
+  public void testChangeServiceType() {
+    // Given
+    Service svc = client.services().inNamespace(session.getNamespace()).withName("service-change-service-type").get();
+
+    // When
+    svc.getSpec().setType("ExternalName");
+    svc.getSpec().setExternalName("my.database.example.com");
+    svc.getSpec().setClusterIP("");
```
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply kubernetes-itests/src/test/java/io/fabric8/kubernetes/ServiceIT.java:107]: Services of type `ExternalName` don't need ClusterIP, API Servers enforces the ClusterIP field to be empty when editing the 'type' field from ClusterIP to ExternalName

File: kubernetes-itests/src/test/java/io/fabric8/kubernetes/ServiceIT.java [hunk 2]
```diff
@@ -94,6 +96,135 @@ public void delete() {
...
+    svc = client.services().inNamespace(session.getNamespace()).createOrReplace(svc);
+
+    // Then
+    assertNotNull(svc);
+    assertEquals("ExternalName", svc.getSpec().getType());
+    assertEquals("my.database.example.com", svc.getSpec().getExternalName());
+  }
+
+  @Test
+  public void testClusterIPCreateOrReplace() {
```
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply kubernetes-itests/src/test/java/io/fabric8/kubernetes/ServiceIT.java:117]: Yes, I'll create a follow up PR with your suggested changes

File: kubernetes-itests/src/test/java/io/fabric8/kubernetes/ServiceIT.java [hunk 3]
```diff
@@ -94,6 +96,135 @@ public void delete() {
...
+    client.services().inNamespace(session.getNamespace()).createOrReplace(clusterIPSvc);
+    // Modify resource
+    clusterIPSvc.getSpec().getPorts().get(0).setTargetPort(new IntOrString(81));
+    // Do createOrReplace again; resource should get updated
+    clusterIPSvc = client.services().inNamespace(session.getNamespace()).createOrReplace(clusterIPSvc);
+
+    // Then
+    assertNotNull(clusterIPSvc);
+    assertEquals("NodePort", clusterIPSvc.getSpec().getType());
+    assertEquals(81, clusterIPSvc.getSpec().getPorts().get(0).getTargetPort().getIntVal().intValue());
```
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply kubernetes-itests/src/test/java/io/fabric8/kubernetes/ServiceIT.java:172]: I added this assertion to check whether what I modified in When phase is actually reflected or not. But yes, I can try doing something with NodePort too

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/core/v1/ServiceOperationsImpl.java [hunk 4]
```diff
@@ -58,30 +60,21 @@ public ServiceOperationsImpl newInstance(OperationContext context) {
...
-        Service old = getMandatory();
-        return super.patch(new ServiceBuilder(item)
-          .editSpec()
-          .withClusterIP(old.getSpec().getClusterIP())
-          .endSpec()
-          .build());
-      } catch (Exception e) {
-        throw KubernetesClientException.launderThrowable(forOperationType("patch"), e);
-      }
+    try {
```
Comment (jorsol) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/core/v1/ServiceOperationsImpl.java:68]: Shouldn't the `patch` also work with `ExternalName`?

Comment (rohanKanojia) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/core/v1/ServiceOperationsImpl.java:68]: I think you're right. Could you please create an issue for this?

**Pull request merged**